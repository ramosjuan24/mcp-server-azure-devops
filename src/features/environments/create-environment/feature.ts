import { WebApi } from 'azure-devops-node-api';
import { CreateEnvironmentOptions } from '../types';
import {
  AzureDevOpsError,
  AzureDevOpsAuthenticationError,
  AzureDevOpsResourceNotFoundError,
} from '../../../shared/errors';

export async function createEnvironment(
  connection: WebApi,
  options: CreateEnvironmentOptions,
) {
  try {
    const releaseApi = await connection.getReleaseApi();

    const environment = {
      name: options.name,
      description: options.description,
    };

    const result = await releaseApi.createEnvironment(
      environment,
      options.projectId,
    );

    return result;
  } catch (error) {
    if (error instanceof Error) {
      if (error.message.includes('Authentication failed')) {
        throw new AzureDevOpsAuthenticationError(
          'Failed to authenticate with Azure DevOps',
        );
      }
      if (error.message.includes('not found')) {
        throw new AzureDevOpsResourceNotFoundError('Project not found');
      }
    }
    throw new AzureDevOpsError('Failed to create environment');
  }
}
