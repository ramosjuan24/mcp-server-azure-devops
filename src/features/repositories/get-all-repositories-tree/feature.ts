import { WebApi } from 'azure-devops-node-api';
import {
  GitVersionType,
  VersionControlRecursionType,
  GitItem,
  GitObjectType,
} from 'azure-devops-node-api/interfaces/GitInterfaces';
import { minimatch } from 'minimatch';
import { AzureDevOpsError } from '../../../shared/errors';
import {
  GetAllRepositoriesTreeOptions,
  AllRepositoriesTreeResponse,
  RepositoryTreeResponse,
  RepositoryTreeItem,
  GitRepository,
} from '../types';

/**
 * Get tree view of files/directories across multiple repositories
 *
 * @param connection The Azure DevOps WebApi connection
 * @param options Options for getting repository tree
 * @returns Tree structure for each repository
 */
export async function getAllRepositoriesTree(
  connection: WebApi,
  options: GetAllRepositoriesTreeOptions,
): Promise<AllRepositoriesTreeResponse> {
  try {
    const gitApi = await connection.getGitApi();
    let repositories: GitRepository[] = [];

    // Get all repositories in the project
    repositories = await gitApi.getRepositories(options.projectId);

    // Filter repositories by name pattern if specified
    if (options.repositoryPattern) {
      repositories = repositories.filter((repo) =>
        minimatch(repo.name || '', options.repositoryPattern || '*'),
      );
    }

    // Initialize results array
    const results: RepositoryTreeResponse[] = [];

    // Process each repository
    for (const repo of repositories) {
      try {
        // Get default branch ref
        const defaultBranch = repo.defaultBranch;
        if (!defaultBranch) {
          // Skip repositories with no default branch
          results.push({
            name: repo.name || 'Unknown',
            tree: [],
            stats: { directories: 0, files: 0 },
            error: 'No default branch found',
          });
          continue;
        }

        // Clean the branch name (remove refs/heads/ prefix)
        const branchRef = defaultBranch.replace('refs/heads/', '');

        // Initialize tree items array and counters
        const treeItems: RepositoryTreeItem[] = [];
        const stats = { directories: 0, files: 0 };

        // Determine the recursion level and processing approach
        const depth = options.depth !== undefined ? options.depth : 0; // Default to 0 (max depth)

        if (depth === 0) {
          // For max depth (0), use server-side recursion for better performance
          const allItems = await gitApi.getItems(
            repo.id || '',
            options.projectId,
            '/',
            VersionControlRecursionType.Full, // Use full recursion
            true,
            false,
            false,
            false,
            {
              version: branchRef,
              versionType: GitVersionType.Branch,
            },
          );

          // Filter out the root item itself and bad items
          const itemsToProcess = allItems.filter(
            (item) =>
              item.path !== '/' && item.gitObjectType !== GitObjectType.Bad,
          );

          // Process all items at once (they're already retrieved recursively)
          processItemsNonRecursive(
            itemsToProcess,
            treeItems,
            stats,
            options.pattern,
          );
        } else {
          // For limited depth, use the regular recursive approach
          // Get items at the root level
          const rootItems = await gitApi.getItems(
            repo.id || '',
            options.projectId,
            '/',
            VersionControlRecursionType.OneLevel,
            true,
            false,
            false,
            false,
            {
              version: branchRef,
              versionType: GitVersionType.Branch,
            },
          );

          // Filter out the root item itself and bad items
          const itemsToProcess = rootItems.filter(
            (item) =>
              item.path !== '/' && item.gitObjectType !== GitObjectType.Bad,
          );

          // Process the root items and their children (up to specified depth)
          await processItems(
            gitApi,
            repo.id || '',
            options.projectId,
            itemsToProcess,
            branchRef,
            treeItems,
            stats,
            1,
            depth,
            options.pattern,
          );
        }

        // Add repository tree to results
        results.push({
          name: repo.name || 'Unknown',
          tree: treeItems,
          stats,
        });
      } catch (repoError) {
        // Handle errors for individual repositories
        results.push({
          name: repo.name || 'Unknown',
          tree: [],
          stats: { directories: 0, files: 0 },
          error: `Error processing repository: ${repoError instanceof Error ? repoError.message : String(repoError)}`,
        });
      }
    }

    return { repositories: results };
  } catch (error) {
    if (error instanceof AzureDevOpsError) {
      throw error;
    }
    throw new Error(
      `Failed to get repository tree: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}

/**
 * Process items non-recursively when they're already retrieved with VersionControlRecursionType.Full
 */
function processItemsNonRecursive(
  items: GitItem[],
  result: RepositoryTreeItem[],
  stats: { directories: number; files: number },
  pattern?: string,
): void {
  // Sort items (folders first, then by path)
  const sortedItems = [...items].sort((a, b) => {
    if (a.isFolder === b.isFolder) {
      return (a.path || '').localeCompare(b.path || '');
    }
    return a.isFolder ? -1 : 1;
  });

  for (const item of sortedItems) {
    const name = item.path?.split('/').pop() || '';
    const path = item.path || '';
    const isFolder = !!item.isFolder;

    // Skip the root folder
    if (path === '/') {
      continue;
    }

    // Calculate level from path segments
    // Remove leading '/' then count segments
    // For paths like:
    // /README.md -> ["README.md"] -> length 1 -> level 1
    // /src/index.ts -> ["src", "index.ts"] -> length 2 -> level 2
    // /src/utils/helper.ts -> ["src", "utils", "helper.ts"] -> length 3 -> level 3
    const pathSegments = path.replace(/^\//, '').split('/');
    const level = pathSegments.length;

    // Filter files based on pattern (if specified)
    if (!isFolder && pattern && !minimatch(name, pattern)) {
      continue;
    }

    // Add item to results
    result.push({
      name,
      path,
      isFolder,
      level,
    });

    // Update counters
    if (isFolder) {
      stats.directories++;
    } else {
      stats.files++;
    }
  }
}

/**
 * Process items recursively up to the specified depth
 */
async function processItems(
  gitApi: any,
  repoId: string,
  projectId: string,
  items: GitItem[],
  branchRef: string,
  result: RepositoryTreeItem[],
  stats: { directories: number; files: number },
  currentDepth: number,
  maxDepth: number,
  pattern?: string,
): Promise<void> {
  // Sort items (directories first, then files)
  const sortedItems = [...items].sort((a, b) => {
    if (a.isFolder === b.isFolder) {
      return (a.path || '').localeCompare(b.path || '');
    }
    return a.isFolder ? -1 : 1;
  });

  for (const item of sortedItems) {
    const name = item.path?.split('/').pop() || '';
    const path = item.path || '';
    const isFolder = !!item.isFolder;

    // Filter files based on pattern (if specified)
    if (!isFolder && pattern && !minimatch(name, pattern)) {
      continue;
    }

    // Add item to results
    result.push({
      name,
      path,
      isFolder,
      level: currentDepth,
    });

    // Update counters
    if (isFolder) {
      stats.directories++;
    } else {
      stats.files++;
    }

    // Recursively process folders if not yet at max depth
    if (isFolder && currentDepth < maxDepth) {
      try {
        const childItems = await gitApi.getItems(
          repoId,
          projectId,
          path,
          VersionControlRecursionType.OneLevel,
          true,
          false,
          false,
          false,
          {
            version: branchRef,
            versionType: GitVersionType.Branch,
          },
        );

        // Filter out the parent folder itself and bad items
        const itemsToProcess = childItems.filter(
          (child: GitItem) =>
            child.path !== path && child.gitObjectType !== GitObjectType.Bad,
        );

        // Process child items
        await processItems(
          gitApi,
          repoId,
          projectId,
          itemsToProcess,
          branchRef,
          result,
          stats,
          currentDepth + 1,
          maxDepth,
          pattern,
        );
      } catch (error) {
        // Ignore errors in child items and continue with siblings
        console.error(`Error processing folder ${path}: ${error}`);
      }
    }
  }
}

/**
 * Convert the tree items to a formatted ASCII string representation
 *
 * @param repoName Repository name
 * @param items Tree items
 * @param stats Statistics about files and directories
 * @returns Formatted ASCII string
 */
export function formatRepositoryTree(
  repoName: string,
  items: RepositoryTreeItem[],
  stats: { directories: number; files: number },
  error?: string,
): string {
  let output = `${repoName}/\n`;

  if (error) {
    output += `  (${error})\n`;
  } else if (items.length === 0) {
    output += '  (Repository is empty or default branch not found)\n';
  } else {
    // Sort items by path to ensure proper sequence
    const sortedItems = [...items].sort((a, b) => {
      // Sort by level first
      if (a.level !== b.level) {
        return a.level - b.level;
      }
      // Then folders before files
      if (a.isFolder !== b.isFolder) {
        return a.isFolder ? -1 : 1;
      }
      // Then alphabetically
      return a.path.localeCompare(b.path);
    });

    // Create a structured tree representation
    const tree = createTreeStructure(sortedItems);

    // Format the tree starting from the root
    output += formatTree(tree, '  ');
  }

  // Add summary line
  output += `${stats.directories} directories, ${stats.files} files\n`;

  return output;
}

/**
 * Create a structured tree from the flat list of items
 */
function createTreeStructure(items: RepositoryTreeItem[]): TreeNode {
  const root: TreeNode = {
    name: '',
    path: '',
    isFolder: true,
    children: [],
  };

  // Map to track all nodes by path
  const nodeMap: Record<string, TreeNode> = { '': root };

  // First create all nodes
  for (const item of items) {
    nodeMap[item.path] = {
      name: item.name,
      path: item.path,
      isFolder: item.isFolder,
      children: [],
    };
  }

  // Then build the hierarchy
  for (const item of items) {
    if (item.path === '/') continue;

    const node = nodeMap[item.path];
    const lastSlashIndex = item.path.lastIndexOf('/');

    // For root level items, the parent path is empty
    const parentPath =
      lastSlashIndex <= 0 ? '' : item.path.substring(0, lastSlashIndex);

    // Get parent node (defaults to root if parent not found)
    const parent = nodeMap[parentPath] || root;

    // Add this node as a child of its parent
    parent.children.push(node);
  }

  return root;
}

/**
 * Format a tree structure into an ASCII tree representation
 */
function formatTree(node: TreeNode, indent: string): string {
  if (!node.children.length) return '';

  let output = '';

  // Sort the children: folders first, then alphabetically
  const children = [...node.children].sort((a, b) => {
    if (a.isFolder !== b.isFolder) {
      return a.isFolder ? -1 : 1;
    }
    return a.name.localeCompare(b.name);
  });

  // Format each child node
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    const isLast = i === children.length - 1;
    const connector = isLast ? '`-- ' : '|-- ';
    const childIndent = isLast ? '    ' : '|   ';

    // Add the node itself
    const suffix = child.isFolder ? '/' : '';
    output += `${indent}${connector}${child.name}${suffix}\n`;

    // Recursively add its children
    if (child.children.length > 0) {
      output += formatTree(child, indent + childIndent);
    }
  }

  return output;
}

/**
 * Tree node interface for hierarchical representation
 */
interface TreeNode {
  name: string;
  path: string;
  isFolder: boolean;
  children: TreeNode[];
}
