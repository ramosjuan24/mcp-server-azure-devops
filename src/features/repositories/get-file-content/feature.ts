import { WebApi } from 'azure-devops-node-api';
import {
  GitVersionDescriptor,
  GitItem,
  GitVersionType,
  VersionControlRecursionType,
} from 'azure-devops-node-api/interfaces/GitInterfaces';
import { AzureDevOpsResourceNotFoundError } from '../../../shared/errors';

/**
 * Response format for file content
 */
export interface FileContentResponse {
  content: string;
  isDirectory: boolean;
}

/**
 * Get content of a file or directory from a repository
 *
 * @param connection - Azure DevOps WebApi connection
 * @param projectId - Project ID or name
 * @param repositoryId - Repository ID or name
 * @param path - Path to file or directory
 * @param versionDescriptor - Optional version descriptor for retrieving file at specific commit/branch/tag
 * @returns Content of the file or list of items if path is a directory
 */
export async function getFileContent(
  connection: WebApi,
  projectId: string,
  repositoryId: string,
  path: string = '/',
  versionDescriptor?: { versionType: GitVersionType; version: string },
): Promise<FileContentResponse> {
  try {
    const gitApi = await connection.getGitApi();

    // Create version descriptor for API requests
    const gitVersionDescriptor: GitVersionDescriptor | undefined =
      versionDescriptor
        ? {
            version: versionDescriptor.version,
            versionType: versionDescriptor.versionType,
            versionOptions: undefined,
          }
        : undefined;

    // First, try to get items using the path to determine if it's a directory
    let isDirectory = false;
    let items: GitItem[] = [];

    try {
      items = await gitApi.getItems(
        repositoryId,
        projectId,
        path,
        VersionControlRecursionType.OneLevel,
        undefined,
        undefined,
        undefined,
        undefined,
        gitVersionDescriptor,
      );

      // If multiple items are returned or the path ends with /, it's a directory
      isDirectory = items.length > 1 || (path !== '/' && path.endsWith('/'));
    } catch {
      // If getItems fails, try to get file content directly
      isDirectory = false;
    }

    if (isDirectory) {
      // For directories, return a formatted list of the items
      return {
        content: JSON.stringify(items, null, 2),
        isDirectory: true,
      };
    } else {
      // For files, get the actual content
      try {
        // Get file content using the Git API
        const contentStream = await gitApi.getItemContent(
          repositoryId,
          path,
          projectId,
          undefined,
          undefined,
          undefined,
          undefined,
          false,
          gitVersionDescriptor,
          true,
        );

        // Convert the stream to a string
        if (contentStream) {
          const chunks: Buffer[] = [];

          // Listen for data events to collect chunks
          contentStream.on('data', (chunk) => {
            chunks.push(Buffer.from(chunk));
          });

          // Use a promise to wait for the stream to finish
          const content = await new Promise<string>((resolve, reject) => {
            contentStream.on('end', () => {
              // Concatenate all chunks and convert to string
              const buffer = Buffer.concat(chunks);
              resolve(buffer.toString('utf8'));
            });

            contentStream.on('error', (err) => {
              reject(err);
            });
          });

          return {
            content,
            isDirectory: false,
          };
        }

        throw new Error('No content returned from API');
      } catch (error) {
        // If it's a 404 or similar error, throw a ResourceNotFoundError
        if (
          error instanceof Error &&
          (error.message.includes('not found') ||
            error.message.includes('does not exist'))
        ) {
          throw new AzureDevOpsResourceNotFoundError(
            `Path '${path}' not found in repository '${repositoryId}' of project '${projectId}'`,
          );
        }
        throw error;
      }
    }
  } catch (error) {
    // If it's already an AzureDevOpsResourceNotFoundError, rethrow it
    if (error instanceof AzureDevOpsResourceNotFoundError) {
      throw error;
    }

    // Otherwise, wrap it in a ResourceNotFoundError
    throw new AzureDevOpsResourceNotFoundError(
      `Failed to get content for path '${path}': ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
