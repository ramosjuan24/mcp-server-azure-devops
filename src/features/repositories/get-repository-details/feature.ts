import { WebApi } from 'azure-devops-node-api';
import { GitVersionType } from 'azure-devops-node-api/interfaces/GitInterfaces';
import {
  AzureDevOpsResourceNotFoundError,
  AzureDevOpsError,
} from '../../../shared/errors';
import { GetRepositoryDetailsOptions, RepositoryDetails } from '../types';

/**
 * Get detailed information about a repository
 *
 * @param connection The Azure DevOps WebApi connection
 * @param options Options for getting repository details
 * @returns The repository details including optional statistics and refs
 * @throws {AzureDevOpsResourceNotFoundError} If the repository is not found
 */
export async function getRepositoryDetails(
  connection: WebApi,
  options: GetRepositoryDetailsOptions,
): Promise<RepositoryDetails> {
  try {
    const gitApi = await connection.getGitApi();

    // Get the basic repository information
    const repository = await gitApi.getRepository(
      options.repositoryId,
      options.projectId,
    );

    if (!repository) {
      throw new AzureDevOpsResourceNotFoundError(
        `Repository '${options.repositoryId}' not found in project '${options.projectId}'`,
      );
    }

    // Initialize the response object
    const response: RepositoryDetails = {
      repository,
    };

    // Get branch statistics if requested
    if (options.includeStatistics) {
      let baseVersionDescriptor = undefined;

      // If a specific branch name is provided, create a version descriptor for it
      if (options.branchName) {
        baseVersionDescriptor = {
          version: options.branchName,
          versionType: GitVersionType.Branch,
        };
      }

      const branchStats = await gitApi.getBranches(
        repository.id || '',
        options.projectId,
        baseVersionDescriptor,
      );

      response.statistics = {
        branches: branchStats || [],
      };
    }

    // Get repository refs if requested
    if (options.includeRefs) {
      const filter = options.refFilter || undefined;
      const refs = await gitApi.getRefs(
        repository.id || '',
        options.projectId,
        filter,
      );

      if (refs) {
        response.refs = {
          value: refs,
          count: refs.length,
        };
      } else {
        response.refs = {
          value: [],
          count: 0,
        };
      }
    }

    return response;
  } catch (error) {
    if (error instanceof AzureDevOpsError) {
      throw error;
    }
    throw new Error(
      `Failed to get repository details: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
