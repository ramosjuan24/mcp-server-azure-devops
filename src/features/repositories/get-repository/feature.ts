import { WebApi } from 'azure-devops-node-api';
import {
  AzureDevOpsResourceNotFoundError,
  AzureDevOpsError,
} from '../../../shared/errors';
import { GitRepository } from '../types';

/**
 * Get a repository by ID or name
 *
 * @param connection The Azure DevOps WebApi connection
 * @param projectId The ID or name of the project
 * @param repositoryId The ID or name of the repository
 * @returns The repository details
 * @throws {AzureDevOpsResourceNotFoundError} If the repository is not found
 */
export async function getRepository(
  connection: WebApi,
  projectId: string,
  repositoryId: string,
): Promise<GitRepository> {
  try {
    const gitApi = await connection.getGitApi();
    const repository = await gitApi.getRepository(repositoryId, projectId);

    if (!repository) {
      throw new AzureDevOpsResourceNotFoundError(
        `Repository '${repositoryId}' not found in project '${projectId}'`,
      );
    }

    return repository;
  } catch (error) {
    if (error instanceof AzureDevOpsError) {
      throw error;
    }
    throw new Error(
      `Failed to get repository: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
