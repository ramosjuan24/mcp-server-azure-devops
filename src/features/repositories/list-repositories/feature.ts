import { WebApi } from 'azure-devops-node-api';
import { AzureDevOpsError } from '../../../shared/errors';
import { ListRepositoriesOptions, GitRepository } from '../types';

/**
 * List repositories in a project
 *
 * @param connection The Azure DevOps WebApi connection
 * @param options Parameters for listing repositories
 * @returns Array of repositories
 */
export async function listRepositories(
  connection: WebApi,
  options: ListRepositoriesOptions,
): Promise<GitRepository[]> {
  try {
    const gitApi = await connection.getGitApi();
    const repositories = await gitApi.getRepositories(
      options.projectId,
      options.includeLinks,
    );

    return repositories;
  } catch (error) {
    if (error instanceof AzureDevOpsError) {
      throw error;
    }
    throw new Error(
      `Failed to list repositories: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
