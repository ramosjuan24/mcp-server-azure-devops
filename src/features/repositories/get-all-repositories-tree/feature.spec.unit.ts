import { WebApi } from 'azure-devops-node-api';
import {
  GitObjectType,
  VersionControlRecursionType,
} from 'azure-devops-node-api/interfaces/GitInterfaces';
import { getAllRepositoriesTree, formatRepositoryTree } from './feature';
import { RepositoryTreeItem } from '../types';

// Mock the Azure DevOps API
jest.mock('azure-devops-node-api');

describe('getAllRepositoriesTree', () => {
  // Sample repositories
  const mockRepos = [
    {
      id: 'repo1-id',
      name: 'repo1',
      defaultBranch: 'refs/heads/main',
    },
    {
      id: 'repo2-id',
      name: 'repo2',
      defaultBranch: 'refs/heads/master',
    },
    {
      id: 'repo3-id',
      name: 'repo3-api',
      defaultBranch: null, // No default branch
    },
  ];

  // Sample files/folders for repo1 at root level
  const mockRepo1RootItems = [
    {
      path: '/',
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/README.md',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src',
      isFolder: true,
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/package.json',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
  ];

  // Sample files/folders for repo1 - src folder
  const mockRepo1SrcItems = [
    {
      path: '/src',
      isFolder: true,
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/src/index.ts',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src/utils',
      isFolder: true,
      gitObjectType: GitObjectType.Tree,
    },
  ];

  // Sample files/folders for repo1 with unlimited depth (what server would return for Full recursion)
  const mockRepo1FullRecursionItems = [
    {
      path: '/',
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/README.md',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src',
      isFolder: true,
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/package.json',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src/index.ts',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src/utils',
      isFolder: true,
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/src/utils/helper.ts',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/src/utils/constants.ts',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
  ];

  // Sample files/folders for repo2
  const mockRepo2RootItems = [
    {
      path: '/',
      gitObjectType: GitObjectType.Tree,
    },
    {
      path: '/README.md',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
    {
      path: '/data.json',
      isFolder: false,
      gitObjectType: GitObjectType.Blob,
    },
  ];

  let mockConnection: jest.Mocked<WebApi>;
  let mockGitApi: any;

  beforeEach(() => {
    // Clear mocks
    jest.clearAllMocks();

    // Create mock GitApi
    mockGitApi = {
      getRepositories: jest.fn().mockResolvedValue(mockRepos),
      getItems: jest
        .fn()
        .mockImplementation((repoId, _projectId, path, recursionLevel) => {
          if (repoId === 'repo1-id') {
            if (recursionLevel === VersionControlRecursionType.Full) {
              return Promise.resolve(mockRepo1FullRecursionItems);
            } else if (path === '/') {
              return Promise.resolve(mockRepo1RootItems);
            } else if (path === '/src') {
              return Promise.resolve(mockRepo1SrcItems);
            }
          } else if (repoId === 'repo2-id') {
            if (recursionLevel === VersionControlRecursionType.Full) {
              return Promise.resolve(mockRepo2RootItems);
            } else if (path === '/') {
              return Promise.resolve(mockRepo2RootItems);
            }
          }
          return Promise.resolve([]);
        }),
    };

    // Create mock connection
    mockConnection = {
      getGitApi: jest.fn().mockResolvedValue(mockGitApi),
    } as unknown as jest.Mocked<WebApi>;
  });

  it('should return tree structures for multiple repositories with limited depth', async () => {
    // Arrange
    const options = {
      organizationId: 'testOrg',
      projectId: 'testProject',
      depth: 2, // Limited depth
    };

    // Act
    const result = await getAllRepositoriesTree(mockConnection, options);

    // Assert
    expect(mockGitApi.getRepositories).toHaveBeenCalledWith('testProject');
    expect(result.repositories.length).toBe(3);

    // Verify repo1 tree
    const repo1 = result.repositories.find((r) => r.name === 'repo1');
    expect(repo1).toBeDefined();
    expect(repo1?.tree.length).toBeGreaterThan(0);
    expect(repo1?.stats.directories).toBeGreaterThan(0);
    expect(repo1?.stats.files).toBeGreaterThan(0);

    // Verify repo2 tree
    const repo2 = result.repositories.find((r) => r.name === 'repo2');
    expect(repo2).toBeDefined();
    expect(repo2?.tree.length).toBeGreaterThan(0);

    // Verify repo3 has error (no default branch)
    const repo3 = result.repositories.find((r) => r.name === 'repo3-api');
    expect(repo3).toBeDefined();
    expect(repo3?.error).toContain('No default branch found');

    // Verify recursion level was set correctly
    expect(mockGitApi.getItems).toHaveBeenCalledWith(
      'repo1-id',
      'testProject',
      '/',
      VersionControlRecursionType.OneLevel,
      expect.anything(),
      expect.anything(),
      expect.anything(),
      expect.anything(),
      expect.anything(),
    );
  });

  it('should return tree structures with max depth using Full recursion', async () => {
    // Arrange
    const options = {
      organizationId: 'testOrg',
      projectId: 'testProject',
      depth: 0, // Max depth
    };

    // Act
    const result = await getAllRepositoriesTree(mockConnection, options);

    // Assert
    expect(mockGitApi.getRepositories).toHaveBeenCalledWith('testProject');
    expect(result.repositories.length).toBe(3);

    // Verify repo1 tree
    const repo1 = result.repositories.find((r) => r.name === 'repo1');
    expect(repo1).toBeDefined();
    expect(repo1?.tree.length).toBeGreaterThan(0);
    // Should include all items, including nested ones
    expect(repo1?.tree.length).toBe(mockRepo1FullRecursionItems.length - 1); // -1 for root folder

    // Verify recursion level was set correctly
    expect(mockGitApi.getItems).toHaveBeenCalledWith(
      'repo1-id',
      'testProject',
      '/',
      VersionControlRecursionType.Full,
      expect.anything(),
      expect.anything(),
      expect.anything(),
      expect.anything(),
      expect.anything(),
    );

    // Verify all levels are represented
    if (repo1) {
      const level1Items = repo1.tree.filter((item) => item.level === 1);
      const level2Items = repo1.tree.filter((item) => item.level === 2);
      const level3Items = repo1.tree.filter((item) => item.level === 3);

      // Verify we have items at level 1
      expect(level1Items.length).toBeGreaterThan(0);

      // Verify we have items at level 2 (src/something)
      expect(level2Items.length).toBeGreaterThan(0);

      // Check for level 3 items if they exist in our mock data
      if (
        mockRepo1FullRecursionItems.some((item) => {
          const pathSegments = item.path.split('/').filter(Boolean);
          return pathSegments.length >= 3;
        })
      ) {
        expect(level3Items.length).toBeGreaterThan(0);
      }
    }
  });

  it('should filter repositories by pattern', async () => {
    // Arrange
    const options = {
      organizationId: 'testOrg',
      projectId: 'testProject',
      repositoryPattern: '*api*',
      depth: 1,
    };

    // Act
    const result = await getAllRepositoriesTree(mockConnection, options);

    // Assert
    expect(mockGitApi.getRepositories).toHaveBeenCalledWith('testProject');
    expect(result.repositories.length).toBe(1);
    expect(result.repositories[0].name).toBe('repo3-api');
  });

  it('should format repository tree correctly', () => {
    // Arrange
    const treeItems: RepositoryTreeItem[] = [
      { name: 'src', path: '/src', isFolder: true, level: 1 },
      { name: 'index.ts', path: '/src/index.ts', isFolder: false, level: 2 },
      { name: 'README.md', path: '/README.md', isFolder: false, level: 1 },
    ];
    const stats = { directories: 1, files: 2 };

    // Act
    const formatted = formatRepositoryTree('test-repo', treeItems, stats);

    // Assert
    expect(formatted).toMatchSnapshot();
  });

  it('should format complex repository tree structures correctly', () => {
    // Arrange
    const treeItems: RepositoryTreeItem[] = [
      // Root level files
      { name: 'README.md', path: '/README.md', isFolder: false, level: 1 },
      {
        name: 'package.json',
        path: '/package.json',
        isFolder: false,
        level: 1,
      },
      { name: '.gitignore', path: '/.gitignore', isFolder: false, level: 1 },

      // Multiple folders at root level
      { name: 'src', path: '/src', isFolder: true, level: 1 },
      { name: 'tests', path: '/tests', isFolder: true, level: 1 },
      { name: 'docs', path: '/docs', isFolder: true, level: 1 },

      // Nested src folder structure
      { name: 'components', path: '/src/components', isFolder: true, level: 2 },
      { name: 'utils', path: '/src/utils', isFolder: true, level: 2 },
      { name: 'index.ts', path: '/src/index.ts', isFolder: false, level: 2 },

      // Deeply nested components
      {
        name: 'Button',
        path: '/src/components/Button',
        isFolder: true,
        level: 3,
      },
      { name: 'Card', path: '/src/components/Card', isFolder: true, level: 3 },
      {
        name: 'Button.tsx',
        path: '/src/components/Button/Button.tsx',
        isFolder: false,
        level: 4,
      },
      {
        name: 'Button.styles.ts',
        path: '/src/components/Button/Button.styles.ts',
        isFolder: false,
        level: 4,
      },
      {
        name: 'Button.test.tsx',
        path: '/src/components/Button/Button.test.tsx',
        isFolder: false,
        level: 4,
      },
      {
        name: 'index.ts',
        path: '/src/components/Button/index.ts',
        isFolder: false,
        level: 4,
      },
      {
        name: 'Card.tsx',
        path: '/src/components/Card/Card.tsx',
        isFolder: false,
        level: 4,
      },

      // Utils with files
      {
        name: 'helpers.ts',
        path: '/src/utils/helpers.ts',
        isFolder: false,
        level: 3,
      },
      {
        name: 'constants.ts',
        path: '/src/utils/constants.ts',
        isFolder: false,
        level: 3,
      },

      // Empty folder
      { name: 'assets', path: '/src/assets', isFolder: true, level: 2 },

      // Files with special characters
      {
        name: 'file-with-dashes.js',
        path: '/src/file-with-dashes.js',
        isFolder: false,
        level: 2,
      },
      {
        name: 'file_with_underscores.js',
        path: '/src/file_with_underscores.js',
        isFolder: false,
        level: 2,
      },

      // Folders in test directory
      { name: 'unit', path: '/tests/unit', isFolder: true, level: 2 },
      {
        name: 'integration',
        path: '/tests/integration',
        isFolder: true,
        level: 2,
      },

      // Files in test directories
      { name: 'setup.js', path: '/tests/setup.js', isFolder: false, level: 2 },
      {
        name: 'example.test.js',
        path: '/tests/unit/example.test.js',
        isFolder: false,
        level: 3,
      },

      // Files in docs
      { name: 'API.md', path: '/docs/API.md', isFolder: false, level: 2 },
      {
        name: 'CONTRIBUTING.md',
        path: '/docs/CONTRIBUTING.md',
        isFolder: false,
        level: 2,
      },
    ];

    const stats = { directories: 10, files: 18 };

    // Act
    const formatted = formatRepositoryTree('complex-repo', treeItems, stats);

    // Assert
    expect(formatted).toMatchSnapshot();
  });

  it('should handle repository errors gracefully', async () => {
    // Arrange
    mockGitApi.getItems = jest.fn().mockRejectedValue(new Error('API error'));

    const options = {
      organizationId: 'testOrg',
      projectId: 'testProject',
      depth: 1,
    };

    // Act
    const result = await getAllRepositoriesTree(mockConnection, options);

    // Assert
    expect(result.repositories.length).toBe(3);
    const repo1 = result.repositories.find((r) => r.name === 'repo1');
    expect(repo1?.error).toBeDefined();
  });
});
