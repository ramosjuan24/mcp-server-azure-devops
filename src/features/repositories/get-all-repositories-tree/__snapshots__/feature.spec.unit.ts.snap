// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllRepositoriesTree should format complex repository tree structures correctly 1`] = `
"complex-repo/
  |-- docs/
  |   |-- API.md
  |   \`-- CONTRIBUTING.md
  |-- src/
  |   |-- assets/
  |   |-- components/
  |   |   |-- Button/
  |   |   |   |-- Button.styles.ts
  |   |   |   |-- Button.test.tsx
  |   |   |   |-- Button.tsx
  |   |   |   \`-- index.ts
  |   |   \`-- Card/
  |   |       \`-- Card.tsx
  |   |-- utils/
  |   |   |-- constants.ts
  |   |   \`-- helpers.ts
  |   |-- file_with_underscores.js
  |   |-- file-with-dashes.js
  |   \`-- index.ts
  |-- tests/
  |   |-- integration/
  |   |-- unit/
  |   |   \`-- example.test.js
  |   \`-- setup.js
  |-- .gitignore
  |-- package.json
  \`-- README.md
10 directories, 18 files
"
`;

exports[`getAllRepositoriesTree should format repository tree correctly 1`] = `
"test-repo/
  |-- src/
  |   \`-- index.ts
  \`-- README.md
1 directories, 2 files
"
`;
